operator/controllers/util.go

```diff
@@ -286,3 +290,107 @@ func recordPausedChange(recorder record.EventRecorder, obj runtime.Object, curre
 		recorder.Event(obj, corev1.EventTypeNormal, reason, message)
 	}
 }
+
+func operationToTaskGroupRequests(c client.Client, o handler.MapObject) []ctrl.Request {
+	var result []ctrl.Request
+
+	operation, ok := o.Object.(*operatorv1.Operation)
+	if !ok {
+		return nil
+	}
+
+	actual, err := listTaskGroupsByLabels(c, operation.Labels)
+	if err != nil {
+		return nil
+	}
+
+	for _, ms := range actual.Items {
+		name := client.ObjectKey{Namespace: ms.Namespace, Name: ms.Name}
+		result = append(result, ctrl.Request{NamespacedName: name})
+	}
+
+	return result
+}
+
+func getOwnerOperation(ctx context.Context, c client.Client, obj metav1.ObjectMeta) (*operatorv1.Operation, error) {
+	//TODO: check for controller ref instead of Operation/Kind
+	for _, ref := range obj.OwnerReferences {
+		if ref.Kind == "Operation" && ref.APIVersion == operatorv1.GroupVersion.String() {
+			operation := &operatorv1.Operation{}
+			key := client.ObjectKey{
+				Namespace: obj.Namespace,
+				Name:      ref.Name,
+			}
+			err := c.Get(ctx, key, operation)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error reading controller ref for %s/%s", obj.Namespace, obj.Name)
+			}
+			return operation, nil
+		}
+	}
+	return nil, nil
+}
+
+type matchingSelector struct {
+	selector labels.Selector
+}
+
+func (m matchingSelector) ApplyToList(opts *client.ListOptions) {
+	opts.LabelSelector = m.selector
+}
+
+func listNodesBySelector(c client.Client, selector *metav1.LabelSelector) (*corev1.NodeList, error) {
+	s, err := metav1.LabelSelectorAsSelector(selector)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert TaskGroup.Spec.NodeSelector to a map")
```

Comment:
technically `s` will end up here as a list of Requirement objects (or a Selector) - a list of maps.

not very important, but should it say "to a Selector"?
[2]

Reply:
done